From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lambdadroid <lambdadroid@gmail.com>
Date: Sun, 13 Aug 2017 15:46:25 +0200
Subject: [PATCH 26/54] ANDROID: brcmfmac: Use monotonic boot time for BSS TSF
 (v2)

Reverts 8e6cffb3b42f "brmc80211: dont use jiffies for BSS TSF".
v2: Use monotonic boot time instead of jiffies to avoid overflow
after 5 minutes and problems after system goes into suspend.

Android uses the TSF as timestamp when scanning for WiFi networks
and discards scan results entirely if the TSF is before the time
the scan started.

Use the monotonic boot time to avoid discarding all scan results
when using brcmfmac on Android.

Fixes: 8e6cffb3b42f "brmc80211: dont use jiffies for BSS TSF"
---
 .../net/wireless/broadcom/brcm80211/brcmfmac/cfg80211.c   | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/cfg80211.c b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/cfg80211.c
index c2b6e5c966d0..200c9adc292e 100644
--- a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/cfg80211.c
+++ b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/cfg80211.c
@@ -3009,11 +3009,12 @@ static s32 brcmf_inform_single_bss(struct brcmf_cfg80211_info *cfg,
 	brcmf_dbg(CONN, "Capability: %X\n", notify_capability);
 	brcmf_dbg(CONN, "Beacon interval: %d\n", notify_interval);
 	brcmf_dbg(CONN, "Signal: %d\n", bss_data.signal);
+	brcmf_dbg(CONN, "Timestamp: %llu\n", bss_data.boottime_ns);
 
 	bss = cfg80211_inform_bss_data(wiphy, &bss_data,
 				       CFG80211_BSS_FTYPE_UNKNOWN,
 				       (const u8 *)bi->BSSID,
-				       0, notify_capability,
+				       bss_data.boottime_ns, notify_capability,
 				       notify_interval, notify_ie,
 				       notify_ielen, GFP_KERNEL);
 
@@ -3072,6 +3073,7 @@ static s32 brcmf_inform_ibss(struct brcmf_cfg80211_info *cfg,
 	u8 *buf = NULL;
 	s32 err = 0;
 	u32 freq;
+	u64 notify_timestamp;
 	u16 notify_capability;
 	u16 notify_interval;
 	u8 *notify_ie;
@@ -3109,6 +3111,7 @@ static s32 brcmf_inform_ibss(struct brcmf_cfg80211_info *cfg,
 	cfg->channel = freq;
 	notify_channel = ieee80211_get_channel(wiphy, freq);
 
+	notify_timestamp = ktime_to_us(ktime_get_boottime());
 	notify_capability = le16_to_cpu(bi->capability);
 	notify_interval = le16_to_cpu(bi->beacon_period);
 	notify_ie = (u8 *)bi + le16_to_cpu(bi->ie_offset);
@@ -3119,9 +3122,10 @@ static s32 brcmf_inform_ibss(struct brcmf_cfg80211_info *cfg,
 	brcmf_dbg(CONN, "capability: %X\n", notify_capability);
 	brcmf_dbg(CONN, "beacon interval: %d\n", notify_interval);
 	brcmf_dbg(CONN, "signal: %d\n", notify_signal);
+	brcmf_dbg(CONN, "timestamp: %llu\n", notify_timestamp);
 
 	bss = cfg80211_inform_bss(wiphy, notify_channel,
-				  CFG80211_BSS_FTYPE_UNKNOWN, bssid, 0,
+				  CFG80211_BSS_FTYPE_UNKNOWN, bssid, notify_timestamp,
 				  notify_capability, notify_interval,
 				  notify_ie, notify_ielen, notify_signal,
 				  GFP_KERNEL);
